### PR TITLE
Provide lookup key to access Flutter assets in the APK

### PR DESCRIPTION
**Pull request by szakarias**

### CODE REVIEW

File: shell/platform/android/io/flutter/view/FlutterMain.java [hunk 1]
```diff
@@ -308,4 +308,30 @@ public static String findAppBundlePath(Context applicationContext) {
...
+    /**
+     * Returns the file name for the given asset.
+     * The returned file name can be used to access the asset in the APK
+     * through the {@link AssetManager} API.
+     *
+     * @param asset the name of the asset. The name can be hierarchical
+     * @return      the file name to be used with {@link AssetManager}
+     */
+    public static String getLookupKeyForAsset(String asset) {
+        return sFlutterAssetsDir + File.separator + asset;
```
Comment (sigurdm) [inline shell/platform/android/io/flutter/view/FlutterMain.java]: Perhaps share logic with fromFlutterAssets above

File: shell/platform/android/io/flutter/view/FlutterMain.java [hunk 2]
```diff
@@ -308,4 +308,30 @@ public static String findAppBundlePath(Context applicationContext) {
...
+    /**
+     * Returns the file name for the given asset which originates from the
+     * specified packageName. The returned file name can be used to access
+     * the asset in the APK through the {@link AssetManager} API.
+     *
+     * @param asset       the name of the asset. The name can be hierarchical
+     * @param packageName the name of the package from which the asset originates
+     * @return            the file name to be used with {@link AssetManager}
+     */
+    public static String getLookupKeyForAsset(String asset, String packageName) {
```
Comment (sigurdm) [inline shell/platform/android/io/flutter/view/FlutterMain.java]: How "official" is the convention with /packages/? It seems AssetImage is doing this calculation itself, and AssetBundle doesn't know anything about it. It seems weird that this hook knows about stuff that AssetBundle doesn't.
Should we have this logic in AssetBundle as well, or in general leave it up to clients to figure out?

File: shell/platform/android/io/flutter/view/FlutterMain.java [hunk 3]
```diff
@@ -308,4 +308,30 @@ public static String findAppBundlePath(Context applicationContext) {
...
+        return sFlutterAssetsDir + File.separator + "packages" + File.separator +
```
Comment (sigurdm) [inline shell/platform/android/io/flutter/view/FlutterMain.java]: Nit: perhaps delegate to getLookupKeyForAsset(String asset)?
```
return getLookupKeyForAsset("packages" + File.separator + packageName + File.separator + asset);
```

**Comment by szakarias** [comment]
cc @mravn-google

**Comment by mravn-google** [comment]
API consistency is good. We may want to add package argument to `AssetBundle`, or we may want to add documentation to allow users of `AssetBundle` and `getLookupKeyForAsset` to construct keys for package-supplied assets themselves.

BTW, I was wondering if there was a better place for the `getLookupKeyForAsset` method. `FlutterMain` seems to me to be more of an internal class that isn't normally used directly by app or plugin developers. The ones that are normally used are `FlutterActivity`, `Flutter(Native)View`, `FlutterPluginRegistrar`, and `FlutterApplication`. Not sure if we want to replicate the method across all these though. I do think that `FlutterPluginRegistrar` should be a one stop shop for plugins.

**Comment by szakarias** [comment]
I agree that we should add the package argument to ```AssetBundle.load```.

With respect to ```FlutterMain```, I've had the same thoughts. I actually had it on ```FlutterPluginRegistrar``` to begin with but I changed it for the following reasons:

- If you are developing an app with platform specific code (not a plugin) then FlutterPluginRegistrar is not available and in this case we already need to use  the ``` findAppBundlePath``` method in ```FlutterMain```.  See for instance the ```flutter_view``` example.

- Logically it also fits in FlutterMain since it already deals with assets and stores the relevant information at a very early point in time. 

We could replicate it in say ```FlutterPluginRegistrar``` and/or ```FlutterView```, but I would still say that it should stay in ```FlutterMain``` and then have the replicated methods call this one.

**Comment by mravn-google** [comment]
Sgtm. Your call :-)

**Pull request merged**